### PR TITLE
Release 4.0.0 work

### PR DESCRIPTION
**Pull request by madpah**
**DRAFT RELEASE for 4.0.0**

Issues included:
- #261 
- #266 
- #292 (supplied by @hakandilek)
- #185 
- #322 
- #340 (required for Python 3.11 stability)
- #346 
- #277 

Items in progress:
_N/A_

Forward Ports from `3.x.x`:
- #328 
- #338 (partial #336)
- #334 
- #330

For consideration:
_NONE - Game over :-)_

Fixes/Resolves:
- #332 

Signed-off-by: Paul Horton <paul.horton@owasp.org>

### CODE REVIEW

File: tests/test_deserialize_xml.py [hunk 1]
```diff
@@ -0,0 +1,709 @@
...
+#
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) OWASP Foundation. All Rights Reserved.
+
+from datetime import datetime
+from os.path import dirname, join
+from typing import cast
+from unittest.mock import Mock, patch
+from uuid import UUID
+from xml.etree import ElementTree
```
[the remark this answers is not in the document]
Comment (madpah) [reply tests/test_deserialize_xml.py:25]: @sonatype-lift ignore

File: tests/test_real_world_examples.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+# limitations under the License.
+#
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) OWASP Foundation. All Rights Reserved.
+import unittest
+from datetime import datetime
+from os.path import dirname, join
+from typing import cast
+from unittest.mock import patch
+from xml.etree import ElementTree
```
[the remark this answers is not in the document]
Comment (madpah) [reply tests/test_real_world_examples.py:23]: @sonatype-lift ignore

File: cyclonedx/model/dependency.py [hunk 3]
```diff
@@ -17,33 +17,97 @@
...
+
 from .bom_ref import BomRef
 
 
+class DependencyDependencies(serializable.BaseHelper):  # type: ignore
+
+    @classmethod
+    def serialize(cls, o: object) -> List[str]:
+        if isinstance(o, SortedSet):
+            return list(map(lambda i: str(i.ref), o))
```
[the remark this answers is not in the document]
Comment (madpah) [reply cyclonedx/model/dependency.py:37]: @sonatype-lift ignore

File: tests/test_deserialize_xml.py [hunk 4]
```diff
@@ -0,0 +1,709 @@
...
+
+    # Helper methods
+    def _validate_xml_bom(self, bom: Bom, schema_version: SchemaVersion, fixture: str) -> None:
+        bom.metadata.timestamp = fixed_date_time()
+        bom.validate()
+
+        if schema_version != LATEST_SUPPORTED_SCHEMA_VERSION:
+            # Rewind the BOM to only have data supported by the SchemaVersion in question
+            outputter = get_instance(bom=bom, output_format=OutputFormat.XML, schema_version=schema_version)
+            bom = cast(Bom, Bom.from_xml(data=ElementTree.fromstring(outputter.output_as_string())))
```
[the remark this answers is not in the document]
Comment (madpah) [reply tests/test_deserialize_xml.py:695]: @sonatype-lift ignore

[the remark this answers is not in the document]
Comment (jkowalleck) [reply tests/test_deserialize_xml.py:695]: we should take case of this, @madpah 

if i would use this lib in a web service where users could upload some XML, i would not have control of the XML.
Therefore this library MUST be safe of XEE atacks. read https://resources.infosecinstitute.com/topic/xml-vulnerabilities-still-attractive-targets-attackers/

[the remark this answers is not in the document]
Comment (madpah) [reply tests/test_deserialize_xml.py:695]: @sonatype-lift unignore

[the remark this answers is not in the document]
Comment (madpah) [reply tests/test_deserialize_xml.py:695]: @jkowalleck - by design, the `cyclonedx-python-lib` does not load the XML (we're only doing that in tests).

The magical `from_xml()` method should be supplied XML data loaded by an external system, which will (and absolutely should) deal with XEE attacks.

Make sense?

[the remark this answers is not in the document]
Comment (madpah) [reply tests/test_deserialize_xml.py:695]: @jkowalleck - by design, the `cyclonedx-python-lib` does not load the XML (we're only doing that in tests).

The magical `from_xml()` method should be supplied XML data loaded by an external system, which will (and absolutely should) deal with XEE attacks.

Make sense?

File: cyclonedx/model/dependency.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (madpah) [reply cyclonedx/model/dependency.py:37]: @sonatype-lift ignore - code is too cool for Lift :-)

**Pull request merged**